service/src/main/java/com/commercetools/pspadapter/payone/notification/NotificationProcessorBase.java

```diff
@@ -53,29 +65,81 @@ public final void processTransactionStatusNotification(final Notification notifi
+            // 1. update payment
+            CompletionStage<Order> fullCompletionStage = getPaymentService().updatePayment(payment, createPaymentUpdates(payment, notification))
+                    // 2. when payment is updated - try to find and updated respective Order.paymentState
+                    .thenComposeAsync(this::tryToUpdateOrderByPayment);
+
+            executeBlocking(fullCompletionStage);
+
         } catch (final io.sphere.sdk.client.ConcurrentModificationException e) {
             throw new java.util.ConcurrentModificationException(e);
         }
     }
 
     /**
+     * Fetch an order with respective payment id and try to update that order.
+     *
+     * @param updatedPayment <b>non-null</b> new updated payment instance.
+     * @return completion stage with nullable updated {@link Order}
+     */
+    private CompletionStage<Order> tryToUpdateOrderByPayment(Payment updatedPayment) {
+        return getOrderService().getOrderByPaymentId(updatedPayment.getId())
+                .thenComposeAsync(order -> updateOrderIfExists(order.orElse(null), updatedPayment));
+    }
+
+    /**
+     * If the {@code order} is not null and it's payment status is different from the new one - update it,
+     * otherwise - skip updating.
+     * <p>
+     * If a new payment status can't be mapped to the order's payment state - the value is left unchanged.
+     *
+     * @param order          <b>nullable</b> order to update state. If <b>null</b> - the update is skipped,
+     *                       but the incident is reported.
+     * @param updatedPayment <b>non-null</b> instance of the updated payment
+     * @return a {@link CompletionStage} with <b>nullable</b> updated {@link Order} instance. If incoming order is null
+     * or new payment state is undefined, or update is not required - the returned stage is completed and contains
+     * the same instance.
+     */
+    private CompletionStage<Order> updateOrderIfExists(Order order, Payment updatedPayment) {
+        if (order != null) {
+            PaymentState newPaymentState = getPaymentToOrderStateMapper().mapPaymentToOrderState(updatedPayment);
+
+            if (newPaymentState == null && updatedPayment.getPaymentStatus() != null) {
+                LOG.warn("Payment [{}] has paymentStatus [{}] which can't be mapped to Order#paymentState. "
+                                + "The order's state remains unchanged.",
+                        updatedPayment.getId(), updatedPayment.getPaymentStatus().getInterfaceCode());
+            }
+
+            // skip update for undefined or unchanged payment state
+            if (newPaymentState != null && !newPaymentState.equals(order.getPaymentState())) {
+                return getOrderService().updateOrderPaymentState(order, newPaymentState);
+            }
+        } else {
+            LOG.error("Payment [{}] is updated, but respective order is not found!", updatedPayment.getId());
```

Comment:
This might be an issue as notification "paid/capture" might arrive before user got redirected back to shop and thus before order creation. In this case order payment status will never change. Different async process based on interface interaction messages and order created messages might be more suitable here.